src/components/UserSettingsPage/UserSettingsPage.tsx

```diff
@@ -36,32 +34,42 @@ export const UserSettingsPage = ({ user, ssrTime }: ExternalPageProps) => {
     const updateSettingsMutation = trpc.user.updateSettings.useMutation();
     const utils = trpc.useContext();
 
-    const [actualUserFields, setActualUserFields] = useState({
+    const defaultValues = useRef<UpdateUser | undefined>({
         name: user.name,
         nickname: user.nickname,
     });
-    const [generalFormChanged, setGeneralFormChanged] = useState(false);
 
     const generalForm = useForm<UpdateUser>({
         resolver: zodResolver(updateUserSchema),
         mode: 'onChange',
         reValidateMode: 'onChange',
         shouldFocusError: true,
-        defaultValues: actualUserFields,
+        defaultValues: defaultValues.current,
     });
 
-    const generalFormValues = generalForm.watch();
-    useEffect(() => {
-        setGeneralFormChanged(!shallowEqual(generalFormValues, actualUserFields));
-    }, [generalFormValues, actualUserFields]);
+    const actualValues = generalForm.watch();
+
+    const isEqual = useMemo(
+        () =>
+            actualValues.name === defaultValues.current?.name &&
+            actualValues.nickname === defaultValues.current?.nickname,
+        [defaultValues, actualValues],
```

Comment:
`defaultValues` will not change cuz it's `ref`, use [`formState.isDirty`](https://react-hook-form.com/docs/useform/formstate)  flag instead of `isEqual` and then you can remove `defaultValues = useRef` as well